executor/aggregate.go

```diff
@@ -977,3 +997,284 @@ func (e *groupChecker) reset() {
 		e.tmpGroupKey = e.tmpGroupKey[:0]
 	}
 }
+
+// vecGroupChecker is used to split a given chunk according to the `group by` expression in a vectorized manner
+// It is usually used for streamAgg
+type vecGroupChecker struct {
+	ctx          sessionctx.Context
+	GroupByItems []expression.Expression
+
+	// groupOffset holds the index of the last row in each group of the current chunk
+	groupOffset []int
+	// groupCount is the number of groups in the current chunk
+	groupCount int
+	// curGroupID is the group obtained by the next call to getNextGroup
+	curGroupID int
+
+	// previousLastGroupKey is the groupKey of the last group of the previous chunk
+	previousLastGroupKey []byte
+	// firstGroupKey and lastGroupKey are used to store the groupKey of the first and last group of the current chunk
+	firstGroupKey []byte
+	lastGroupKey  []byte
+
+	// firstRowDatums and lastRowDatums store the results of the expression evaluation for the first and last rows of the current chunk in datum
+	// They are used to encode to get firstGroupKey and lastGroupKey
+	firstRowDatums []types.Datum
+	lastRowDatums  []types.Datum
+
+	// sameGroup is used to check whether two adjacent rows belong to the same group
+	sameGroup []bool
+}
+
+func newVecGroupChecker(ctx sessionctx.Context, items []expression.Expression) *vecGroupChecker {
+	sameGroup := make([]bool, 1024)
+	return &vecGroupChecker{
+		ctx:          ctx,
+		GroupByItems: items,
+		groupCount:   0,
+		curGroupID:   0,
+		sameGroup:    sameGroup,
+	}
+}
+
+// splitIntoGroups divides the chunk into more groups which the row in the same group have the same groupKey
+// the return value areSameGroup means whether the groupKey of the first group of the newly passed chunk is equal to the groupKey of the last group left before
+// TODO: Since all the group by items are only a column reference, guaranteed by building projection below aggregation, we can directly compare data in a chunk.
+func (e *vecGroupChecker) splitIntoGroups(chk *chunk.Chunk) (areSameGroup bool, err error) {
+	// The numRows can not be zero. `FetchChild` is called before `splitIntoGroups` is called.
+	// if numRows == 0, it will be returned in `fetchChild`. See `fetchChild` for more details.
+	numRows := chk.NumRows()
+
+	e.reset()
+	e.curGroupID = 0
+	if len(e.GroupByItems) == 0 {
```

Comment:
Can we define `e.reset` as `e.reset(numRows)`, and move this code block into it?

Reply:
I think it's not better. For the first reason, the ```reset``` also use [here](https://github.com/pingcap/tidb/blob/445e30204a463b8a370e950829c34d6fd16cd4d6/executor/aggregate.go#L811). Besides when the ```if len(e.GroupByItems) == 0``` , the chunk has only one group. It will return. When we put them to ``reset()```, it is not easy to deal with this situation.